File: android/src/main/kotlin/dev/pam/backgroundtransfer/TransferHeaders.kt
package dev.pam.backgroundtransfer

import org.json.JSONObject
import java.util.Locale

internal object TransferHeaders {
    private val reserved = setOf("host", "content-length", "transfer-encoding", "connection", "trailer", "upgrade")
    private val namePattern = Regex("^[A-Za-z0-9!#$%&'*+.^_`|~-]+$")

    fun decode(encoded: String): Map<String, String> {
        require(encoded.toByteArray(Charsets.UTF_8).size <= 4096) { "Transfer headers too large" }
        val json = JSONObject(encoded)
        require(json.length() <= 32) { "Too many transfer headers" }
        val headers = mutableMapOf<String, String>()
        json.keys().forEach { name ->
            val value = json.get(name)
            require(value is String) { "Invalid transfer header" }
            headers[name] = value
        }
        validate(headers)
        return headers
    }

    fun validate(headers: Map<String, String>) {
        require(headers.size <= 32) { "Too many transfer headers" }
        val seen = mutableSetOf<String>()
        headers.forEach { (name, value) ->
            val key = name.lowercase(Locale.ROOT)
            require(namePattern.matches(name) && seen.add(key) && key !in reserved) { "Invalid transfer header name" }
            require(value.none { it.code < 32 || it.code == 127 }) { "Invalid transfer header value" }
        }
    }
}

File: src/TransferHeaders.php
<?php

declare(strict_types=1);

namespace Pam\Native\BackgroundTransfer;

use InvalidArgumentException;

final class TransferHeaders
{
    /** @param array<string, string> $headers */
    public static function encode(array $headers): string
    {
        if (count($headers) > 32) throw new InvalidArgumentException('Too many transfer headers');
        $seen = [];
        foreach ($headers as $name => $value) {
            if (!is_string($name) || preg_match('/^[A-Za-z0-9!#$%&\x27*+.^_`|~-]+$/D', $name) !== 1
                || !is_string($value) || preg_match('/[\x00-\x1f\x7f]/', $value) === 1) {
                throw new InvalidArgumentException('Invalid transfer header');
            }
            $key = strtolower($name);
            if (isset($seen[$key]) || in_array($key, ['host', 'content-length', 'transfer-encoding', 'connection', 'trailer', 'upgrade'], true)) {
                throw new InvalidArgumentException('Reserved or duplicate transfer header');
            }
            $seen[$key] = true;
        }
        $encoded = json_encode((object) $headers, JSON_THROW_ON_ERROR);
        if (strlen($encoded) > 4096) throw new InvalidArgumentException('Transfer headers exceed four KiB');

        return $encoded;
    }
}
